cumulus/parachains/runtimes/assets/common/src/erc20_transactor.rs

```diff
@@ -0,0 +1,195 @@
+
+impl<
+		AccountId: Eq + Clone,
+		T: pallet_revive::Config<AccountId = AccountId>,
+		AccountIdConverter: ConvertLocation<AccountId>,
+		Matcher: MatchesFungibles<H160, u128>,
+		GasLimit: Get<Weight>,
+		TransfersCheckingAccount: Get<AccountId>,
+	> TransactAsset
+	for ERC20Transactor<T, Matcher, AccountIdConverter, GasLimit, AccountId, TransfersCheckingAccount>
+where
+	BalanceOf<T>: Into<U256> + TryFrom<U256>,
+	MomentOf<T>: Into<U256>,
+	T::Hash: frame_support::traits::IsType<H256>,
+{
+	fn can_check_in(_origin: &Location, _what: &Asset, _context: &XcmContext) -> XcmResult {
+		// We don't support teleports.
+		Err(XcmError::Unimplemented)
+	}
+
+	fn check_in(_origin: &Location, _what: &Asset, _context: &XcmContext) {
+		// We don't support teleports.
+	}
+
+	fn can_check_out(_destination: &Location, _what: &Asset, _context: &XcmContext) -> XcmResult {
+		// We don't support teleports.
+		Err(XcmError::Unimplemented)
+	}
+
+	fn check_out(_destination: &Location, _what: &Asset, _context: &XcmContext) {
+		// We don't support teleports.
+	}
+
+	fn withdraw_asset_with_surplus(
+		what: &Asset,
+		who: &Location,
+		_context: Option<&XcmContext>,
+	) -> Result<(AssetsInHolding, Weight), XcmError> {
+		tracing::trace!(
+			target: "xcm::transactor::erc20::withdraw",
+			?what, ?who,
+		);
+		let (asset_id, amount) = Matcher::matches_fungibles(what)?;
+		let who = AccountIdConverter::convert_location(who)
+			.ok_or(MatchError::AccountIdConversionFailed)?;
+		// We need to map the 32 byte checking account to a 20 byte account.
+		let checking_account_eth = T::AddressMapper::to_address(&TransfersCheckingAccount::get());
+		let checking_address = Address::from(Into::<[u8; 20]>::into(checking_account_eth));
+		let gas_limit = GasLimit::get();
+		// To withdraw, we actually transfer to the checking account.
+		// We do this using the solidity ERC20 interface.
+		let data =
+			IERC20::transferCall { to: checking_address, value: EU256::from(amount) }.abi_encode();
+		let ContractResult { result, gas_consumed, .. } = pallet_revive::Pallet::<T>::bare_call(
+			T::RuntimeOrigin::signed(who.clone()),
+			asset_id,
+			BalanceOf::<T>::zero(),
+			gas_limit,
+			DepositLimit::Unchecked,
+			data,
+		);
+		// We need to return this surplus for the executor to allow refunding it.
+		let surplus = gas_limit.saturating_sub(gas_consumed);
+		tracing::trace!(target: "xcm::transactor::erc20::withdraw", ?gas_consumed, ?surplus, "Gas consumed");
+		if let Ok(return_value) = result {
+			tracing::trace!(target: "xcm::transactor::erc20::withdraw", ?return_value, "Return value by withdraw_asset");
+			if return_value.did_revert() {
+				tracing::debug!(target: "xcm::transactor::erc20::withdraw", "ERC20 contract reverted");
+				Err(XcmError::FailedToTransactAsset("ERC20 contract reverted"))
+			} else {
+				let is_success = bool::abi_decode(&return_value.data, false).map_err(|error| {
```

Comment:
I hope this is safe. The contract is free to return as much data as it wants. But it would be a really stupid impl if alloy core did try to decode all the data instead of only what is needed for bool.

Reply:
It seems like it [takes one word at a time](https://github.com/alloy-rs/core/blob/b20e2326796827cbc5ca9ff7bd037fab9ba37e93/crates/sol-types/src/abi/decoder.rs#L189) so for a bool it would be 1 word which are 32 bytes. It shouldn't grab more than one word